surfacers/bigquery/bigquery.go

```diff
@@ -0,0 +1,183 @@
+
+	// Cloud logger
+	l *logger.Logger
+}
+
+// New initializes a bigquery surfacer. bigquery surfacer inserts probe results
+// into a bigquery database.
+// ctx is used to manage background goroutine.
+func New(ctx context.Context, config *configpb.SurfacerConf, l *logger.Logger) (*Surfacer, error) {
+	s := &Surfacer{
+		c: config,
+		l: l,
+	}
+	return s, s.init(ctx)
+}
+
+func (row *bqrow) Save() (map[string]bigquery.Value, string, error) {
+	return row.value, "", nil
+}
+
+// Write takes the data to be written
+func (s *Surfacer) Write(ctx context.Context, em *metrics.EventMetrics) {
+	select {
+	case s.writeChan <- em:
+	default:
+		s.l.Errorf("Surfacer's write channel is full, dropping new data.")
+	}
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func convertToBqType(colType, label string) (bigquery.Value, error) {
+	colType = strings.ToLower(colType)
+	switch colType {
+	case "string":
+		return label, nil
+	case "integer", "int":
+		val, err := strconv.ParseInt(label, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		return val, nil
+	case "float", "double":
+		val, err := strconv.ParseFloat(label, 64)
+		if err != nil {
+			return nil, err
+		}
+		return val, nil
+	case "timestamp":
+		milliSec, err := strconv.ParseInt(label, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		timestamp := time.UnixMilli(milliSec).UTC()
+		return timestamp, nil
+	default:
+		return nil, errors.New("invalid column type")
+	}
+}
+
+func (s *Surfacer) parseBQCols(em *metrics.EventMetrics) (map[string]bigquery.Value, error) {
+	bqRowMap := make(map[string]bigquery.Value)
+	for _, col := range s.c.GetColumns() {
+		colName := col.GetName()
+		val, err := convertToBqType(col.GetType(), em.Label(colName))
+		if err != nil {
+			return nil, fmt.Errorf("error occurred while parsing for field %v: %v", colName, err)
+		}
+		bqRowMap[colName] = val
+	}
+	return bqRowMap, nil
+}
+
+func (s *Surfacer) batchInsertRowsToBQ(ctx context.Context, inserter iInserter) {
+	chanLen := len(s.writeChan)
+	bigqueryTimeout := time.Duration(s.c.GetBigqueryTimeout()) * time.Second
+	bqctx := context.Background()
+	bqctx, cancel := context.WithTimeout(bqctx, bigqueryTimeout)
+	defer cancel()
+
+	for i := 0; i < chanLen; i += batchSize {
+		var results []*bqrow
+
+		for j := i; j < min(i+batchSize, chanLen); j++ {
+			em := <-s.writeChan
+
+			bqRowMap, err := s.parseBQCols(em)
+
+			if err == nil {
+				result := &bqrow{value: bqRowMap}
+				results = append(results, result)
+			} else {
+				s.l.Errorf("%v", err)
+			}
+		}
+
+		if len(results) > 0 {
+			if err := inserter.Put(bqctx, results); err != nil {
+				s.l.Errorf("failed uploading probe results to Bigquery due to surfacer: %v", err)
+			}
+		}
+	}
+}
+
+func (s *Surfacer) writeToBQ(ctx context.Context, inserter iInserter) {
+	bigqueryInsertionTime := s.c.GetBatchInsertionTime()
+	ticker := time.NewTicker(time.Duration(bigqueryInsertionTime) * time.Second)
```

Comment:
Let's add a defer statement to stop this ticket:
```suggestion
	ticker := time.NewTicker(time.Duration(bigqueryInsertionTime) * time.Second)
	defer ticket.Stop()
```